toolchain/sem_ir/name_scope.h

```diff
@@ -18,52 +18,133 @@ enum class AccessKind : int8_t {
   Private,
 };
 
-struct NameScope : Printable<NameScope> {
+class NameScope : public Printable<NameScope> {
+ public:
   struct Entry {
     NameId name_id;
     InstId inst_id;
     AccessKind access_kind;
   };
 
+  struct EntryId : public IdBase {
+    using IdBase::IdBase;
+  };
+
+  explicit NameScope(InstId inst_id, NameId name_id,
+                     NameScopeId parent_scope_id)
+      : inst_id_(inst_id),
+        name_id_(name_id),
+        parent_scope_id_(parent_scope_id) {}
+
   auto Print(llvm::raw_ostream& out) const -> void {
-    out << "{inst: " << inst_id << ", parent_scope: " << parent_scope_id
-        << ", has_error: " << (has_error ? "true" : "false");
+    out << "{inst: " << inst_id() << ", parent_scope: " << parent_scope_id()
+        << ", has_error: " << (has_error_ ? "true" : "false");
 
     out << ", extended_scopes: [";
     llvm::ListSeparator scope_sep;
-    for (auto id : extended_scopes) {
+    for (auto id : extended_scopes()) {
       out << scope_sep << id;
     }
     out << "]";
 
     out << ", names: {";
     llvm::ListSeparator sep;
-    for (auto entry : names) {
+    for (auto entry : entries()) {
       out << sep << entry.name_id << ": " << entry.inst_id;
     }
     out << "}";
 
     out << "}";
   }
 
-  // Adds a name to the scope that must not already exist.
-  auto AddRequired(Entry name_entry) -> void {
-    auto add_name = [&] {
-      int index = names.size();
-      names.push_back(name_entry);
-      return index;
-    };
-    auto result = name_map.Insert(name_entry.name_id, add_name);
-    CARBON_CHECK(result.is_inserted(), "Failed to add required name: {0}",
-                 name_entry.name_id);
+  // Names in the scope.
+  // Entries could become invalidated if the scope object is invalidated or if a
+  // name is added.
+  auto entries() const -> llvm::ArrayRef<Entry> { return names_; }
```

Comment:
Minor, but you might consider clustering simple accessors/mutators below the regular functions. But I think you mostly cluster by functionality right now, which seems fair too.

Reply:
Yes, I've tried both and it seems easier to understand the API if it's groped by functionality.